### DISCUSSION TITLE
Group Sync via the Directory Connector does not work

### DISCUSSION THREAD

**Discussion by bokkabonga** [dani-garcia/vaultwarden#1410] [Q&A]
### Subject of the issue
Syncing Groups via the Directory Connector does not work, while Syncinc Users works just fine. When using the "Test Sync" button the Directory Connector will find the Group and the Users inside the Group but after hitting the "Sync Now" Button no Users will be invited.

  ### Your environment (Generated via diagnostics page)
  * Bitwarden_rs version: v1.19.0-c836f88f
  * Web-vault version: v2.18.1
  * Running within Docker: true
  * Internet access: true
  * Uses a proxy: false
  * DNS Check: true
  * Time Check: true
  * Domain Configuration Check: true
  * HTTPS Check: true
  * Database type: MySQL
  * Database version: MariaDB 10.4
  * Clients used: Directory Connector Portable Windows
  * Reverse proxy and version: 
  * Other relevant information: 

### Steps to reproduce
1. Open Directory Connector and configure it to Connect to LDAP Server. 
2. Define Group Filter and test the Sync
3. Try to actually Sync

### Expected behaviour
Users should be invited to the Organisation.

### Actual behaviour
No Invitations are beeing send. Nothing appears in the Log or the Developer Console of the Directory Connector itself.

**Comment by BlackDex** [comment]
Bitwarden_RS doesn't support groups. Therefor i think that that part of the directory connector will not work with our implementation.

**Reply by bokkabonga** [reply]
Ah i see. I thought that by groups in the directory Connector it meant LDAP Groups, so you could sync a whole LDAP Group with your Organisation. But it is what it is, if it doesn´t work, than thats ok :D

**Reply by BlackDex** [reply]
I think it also means that. But not sure on how this interacts with the server. If there we need to support groups to have this working, then i can understand it will not work. Also, this directory connector part is kinda alpha/beta stage implementation.

**Reply by bokkabonga** [reply]
I would love to jump in on this myself, but my knowledge on the whole project and Rust in particular is...lets say slim :D
I guess when can close this issue for now. (Can i do this myself? Still didn´t find out how to do this, if I can :D )

**Reply by BlackDex** [reply]
You can't as this is a `discussion` and not an issue ;)

**Reply by bokkabonga** [reply]
Smarter everyday :D

**Comment by deajan** [comment]
Since vaultwarden now supports groups, does the directory connector now properly sync groups ?

**Reply by BlackDex** [reply]
I don't think that part was implemented.
Also, we do not even support the latest version of the directory connector.

**Reply by deajan** [reply]
Thank you for your swift answer.
Is there any solution as of today to sync LDAP accounts only without any SSO involved ?
I'd like to find an automagical way to get AD disabled accounts removed / disabled Vaultwarden side.
If the official directory connecter is "the way", what version should I use ?

**Reply by BlackDex** [reply]
I would suggest to checkout https://github.com/ViViDboarder/vaultwarden_ldap
Though, not sure if that it also disables users. I think it only invites people.

**Reply by deajan** [reply]
I actually did google around and found this before posting here.

As far as I read, it claims `This is NOT a sync tool` so it will create new users but not update their state / disable them.
The config file has `ldap_sync_loop` option which is set to True by default, but it's only used to create new users, confirmed by https://github.com/ViViDboarder/vaultwarden_ldap/issues/71

Also, in https://github.com/ViViDboarder/vaultwarden_ldap/issues/77 author says he won't develop anymore new features since vaultwarden is now compatible with official directory connector.

So, if I understand right, `vaultwarden_ldap` is not developped anymore, and official bitwarden directory connector only has partial support.
LDAP support isn't going to exist soon AFAIK.

I'd love help sponsor the directory connector support since I really enjoy this project for myself, my family and my team.
I can put in about 500€, I know this is peanuts for development, and hope other people would do the same. Do you guys have any bounty program or so ?

**Reply by BlackDex** [reply]
Yea i noticed that too after i posted the link to it.
I was working on getting the new version of the directory connector working, that needs an org api-key support.
https://github.com/BlackDex/vaultwarden/tree/org-api-key
Haven't gotten around to it yet to pick it up. There also is a PR which has some support, but not fully tested yet. That one also is lacking behind in merging, so i tended to try and merge that PR into my latest branch and see if i can make it work, but not yet gotten around to it.

**Reply by deajan** [reply]
Well, if you happen to get this done in the next months, just ping me and I'll happily oblige with the sponsoring 8-D

**Reply by Fmstrat** [reply]
Is directory connector 2.9.2 working at all? I got it to say `Synced 7 group(s) and 4 user(s).` but it didn't appear to actually do anything. No users show up, no invites go out, no groups in the org.

**Reply by MatthewA1** [reply]
I'm having the same issue with Directory Connector 2.9.2. Both the GUI and CLI versions report a successful sync. I tried both with group syncing on and off and get the same result (obviously with 0 groups synced when group sync is off).
I set the log level to debug, but it doesn't log anything beyond that the response code:
`POST /api/organizations/<org_id>/import => 200 OK`

**Reply by BlackDex** [reply]
I have created a new PR #3568 , is which should bring the correct support for this. I was able to sync groups and users and it acted upon changes to the users.

I only tested an LDAP environment, but in the end the communication between the Directory Connector and Vaultwarden should stay the same.

**Reply by MatthewA1** [reply]
I just tested this with AD and the latest release of BWDC (2022.11.0), and it seems to work for the most part.
The only major issue I've found so far is that updating a group's Collections access deletes the group's External ID. If BWDC's cache is cleared or the group is otherwise modified, on the next sync, a duplicate group is created as the existing group no longer has the External ID field set.
Edit: Actually, editing the group from the web vault in any way, to include it's membership, causes the External ID to be cleared.

**Reply by BlackDex** [reply]
@MatthewA1 That is also fixed btw in the current `testing`

**Reply by deajan** [reply]
@BlackDex I talked earlier about the sponsoring. Will Liberapay give me a receipt for my accountant ? Thank you again for your work.

**Reply by BlackDex** [reply]
@deajan According to the donation page a receipt is automatically available, so i think it is.

**Reply by deajan** [reply]
@BlackDex Done ;) Thank you for your work.

**Reply by BlackDex** [reply]
@deajan Thanks, Much appreciated!

**Reply by deajan** [reply]
@BlackDex The least I could do. It's been nice to be able to use Vaultwarden for my company ;)
